<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Daniel Greene" />

<meta name="date" content="2024-02-20" />

<title>gsEasy</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>
<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>



<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">gsEasy</h1>
<h4 class="author">Daniel Greene</h4>
<h4 class="date">2024-02-20</h4>



<div id="calculate-p-values-for-enrichment-of-set" class="section level2">
<h2>Calculate <em>p</em>-values for enrichment of set</h2>
<p><code>gsEasy</code> has a function <code>gset</code> for calculating <em>p</em>-values of enrichment for sets (of genes) in ranked/scored lists (of genes) by permutation (see ‘Gene Set Enrichment Analysis’ described by Subramanian et al, 2005). The arguments of <code>gset</code> are named as in the paper:</p>
<ul>
<li><code>N</code>: the total number of genes,</li>
<li><code>S</code>: <code>integer</code> vector giving the ranks of the genes in the test set amongst the <code>N</code>, <em>or</em> giving the indices within the scores vector <code>r</code> (see below) <em>or</em> a <code>character</code> vector of the names of genes in the test set,</li>
<li><code>r</code>: (optional) vector of length <code>N</code> of correlation scores, e.g. gene expression correlation. If unspecified, it defaults to <code>1-(i-1)/N</code> for the <code>i</code>th gene. If <code>S</code> is given as the names of genes, <code>r</code> must either be a <code>character</code> vector of genes in rank order or named by <code>genes</code> (necessarily containing all the genes in <code>S</code>).</li>
<li><code>p</code>: a numeric value used to exponentiate the enrichment scores given by <code>r</code>, with higher values having the effect of increasing the weight on the highest scores/ranks (for more details, see Subramanian et al, 2005). The default value is <code>1</code> [i.e. not transformed].</li>
</ul>
<p>Say we had a set of 5 genes which appeared at the top five ranks out of 1000 (i.e. highly enriched at the high ranks!). We could then calculate an enrichment <em>p</em>-value using the command:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true"></a><span class="kw">gset</span>(<span class="dt">S=</span><span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dt">N=</span><span class="dv">1000</span>)</span></code></pre></div>
<pre><code>## [1] 9.9999e-06</code></pre>
<p>So the <em>p</em>-value is close to zero. However for random sets, the <em>p</em>-values are distributed uniformly:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true"></a><span class="kw">replicate</span>(<span class="dt">n=</span><span class="dv">10</span>, <span class="dt">expr=</span><span class="kw">gset</span>(<span class="dt">S=</span><span class="kw">sample.int</span>(<span class="dt">n=</span><span class="dv">1000</span>, <span class="dt">size=</span><span class="dv">5</span>), <span class="dt">N=</span><span class="dv">1000</span>))</span></code></pre></div>
<pre><code>##  [1] 0.7213930 0.1111111 0.1058394 0.5870647 0.1142857 0.9850746 0.9701493
##  [8] 0.1840796 0.5074627 0.8905473</code></pre>
<p>Alternatively, you can pass the names of genes as <code>S</code> with a sorted list of gene names as <code>r</code> (in which case the scores default to the ranks in the list), or a numeric vector of scores named by genes as <code>r</code>.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true"></a><span class="kw">gset</span>(<span class="dt">S=</span><span class="kw">c</span>(<span class="st">&quot;gene 1&quot;</span>, <span class="st">&quot;gene 5&quot;</span>, <span class="st">&quot;gene 40&quot;</span>), <span class="dt">r=</span><span class="kw">paste</span>(<span class="st">&quot;gene&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">100</span>))</span></code></pre></div>
<pre><code>## [1] 0.07792208</code></pre>
<p>Multiple gene sets can thus be tested for enrichment with a single call to a high level function such as <code>sapply</code> (or, if you have many sets to test and multiple cores available, <code>mclapply</code>), for instance:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true"></a>gene_sets &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="kw">list</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>), <span class="kw">replicate</span>(<span class="dt">n=</span><span class="dv">10</span>, <span class="dt">simplify=</span><span class="ot">FALSE</span>, <span class="dt">expr=</span><span class="kw">sample.int</span>(<span class="dt">n=</span><span class="dv">1000</span>, <span class="dt">size=</span><span class="dv">5</span>)))</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true"></a><span class="kw">names</span>(gene_sets) &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;enriched set&quot;</span>, <span class="kw">paste</span>(<span class="st">&quot;unenriched set&quot;</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>))</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true"></a>gene_sets</span></code></pre></div>
<pre><code>## $`enriched set`
## [1] 1 2 3 4 5
## 
## $`unenriched set 1`
## [1]   6  53 974 143 936
## 
## $`unenriched set 2`
## [1] 438 237 252 458 293
## 
## $`unenriched set 3`
## [1] 624 680 778 230 727
## 
## $`unenriched set 4`
## [1] 997 254 714 307 513
## 
## $`unenriched set 5`
## [1] 104 786 263 857 632
## 
## $`unenriched set 6`
## [1] 149 970 961  73 265
## 
## $`unenriched set 7`
## [1] 435 138 982 387 722
## 
## $`unenriched set 8`
## [1] 667 441 506 186 913
## 
## $`unenriched set 9`
## [1] 791 911 638 468 274
## 
## $`unenriched set 10`
## [1] 758 626 817 698  45</code></pre>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true"></a><span class="kw">sapply</span>(gene_sets, <span class="cf">function</span>(set) <span class="kw">gset</span>(<span class="dt">S=</span>set, <span class="dt">N=</span><span class="dv">1000</span>))</span></code></pre></div>
<pre><code>##      enriched set  unenriched set 1  unenriched set 2  unenriched set 3 
##      0.0000099999      0.0017199828      0.2388059701      0.9353233831 
##  unenriched set 4  unenriched set 5  unenriched set 6  unenriched set 7 
##      0.7562189055      0.5024875622      0.0247097529      0.4228855721 
##  unenriched set 8  unenriched set 9 unenriched set 10 
##      0.7462686567      0.9054726368      0.5124378109</code></pre>
</div>
<div id="ontological-annotations" class="section level2">
<h2>Ontological annotations</h2>
<p><code>gsEasy</code> has a function <code>get_ontological_gene_sets</code> for creating lists of gene sets corresponding to annotation with ontological terms such that ontological <em>is-a</em> relations are propagated. <code>get_ontological_gene_sets</code> accepts an <code>ontological_index</code> (see the R package <code>ontologyIndex</code> for more details) argument and two character vectors, corresponding to genes and terms respectively, whereby the n-th element in each vector corresponds to one annotation pair. The result, a list of character vectors of gene names, can then be used as an argument of <code>gset</code>.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true"></a><span class="kw">library</span>(ontologyIndex)</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true"></a><span class="kw">data</span>(hpo)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true"></a>df &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true"></a>    <span class="dt">gene=</span><span class="kw">c</span>(<span class="st">&quot;gene 1&quot;</span>, <span class="st">&quot;gene 2&quot;</span>), </span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true"></a>    <span class="dt">term=</span><span class="kw">c</span>(<span class="st">&quot;HP:0000598&quot;</span>, <span class="st">&quot;HP:0000118&quot;</span>), </span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true"></a>    <span class="dt">name=</span>hpo<span class="op">$</span>name[<span class="kw">c</span>(<span class="st">&quot;HP:0000598&quot;</span>, <span class="st">&quot;HP:0000118&quot;</span>)], </span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true"></a>    <span class="dt">stringsAsFactors=</span><span class="ot">FALSE</span>,</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true"></a>    <span class="dt">row.names=</span><span class="ot">NULL</span>)</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true"></a>df</span></code></pre></div>
<pre><code>##     gene       term                   name
## 1 gene 1 HP:0000598 Abnormality of the ear
## 2 gene 2 HP:0000118 Phenotypic abnormality</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true"></a><span class="kw">get_ontological_gene_sets</span>(hpo, <span class="dt">gene=</span>df<span class="op">$</span>gene, <span class="dt">term=</span>df<span class="op">$</span>term)</span></code></pre></div>
<pre><code>## $`HP:0000001`
## [1] &quot;gene 1&quot; &quot;gene 2&quot;
## 
## $`HP:0000118`
## [1] &quot;gene 1&quot; &quot;gene 2&quot;
## 
## $`HP:0000598`
## [1] &quot;gene 1&quot;</code></pre>
</div>
<div id="gene-ontology-go-annotations" class="section level2">
<h2>Gene Ontology (GO) annotations</h2>
<p><code>gsEasy</code> comes with a <code>list</code> of GO annotations, <code>GO_gene_sets</code> [based on annotations downloaded from geneontology.org on 07/08/2016], which can be loaded with <code>data</code>. This comprises a <code>list</code> of all gene sets (i.e. <code>character</code> vectors of gene names) associated with each GO term, for GO terms being annotated with at most 500 genes.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true"></a><span class="kw">data</span>(GO_gene_sets)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true"></a>GO_gene_sets[<span class="dv">1</span><span class="op">:</span><span class="dv">6</span>]</span></code></pre></div>
<pre><code>## $`GO:0000002`
##  [1] &quot;AKT3&quot;     &quot;LONP1&quot;    &quot;MEF2A&quot;    &quot;MGME1&quot;    &quot;MPV17&quot;    &quot;MRPL17&quot;  
##  [7] &quot;MRPL39&quot;   &quot;OPA1&quot;     &quot;PIF1&quot;     &quot;SLC25A33&quot; &quot;SLC25A36&quot; &quot;SLC25A4&quot; 
## [13] &quot;TYMP&quot;    
## 
## $`GO:0000003`
##  [1] &quot;EIF4H&quot;  &quot;IL12B&quot;  &quot;LEP&quot;    &quot;LEPR&quot;   &quot;MMP23A&quot; &quot;RHOXF1&quot; &quot;SEPP1&quot;  &quot;STAT3&quot; 
##  [9] &quot;TNP1&quot;   &quot;VGF&quot;    &quot;WDR43&quot; 
## 
## $`GO:0000009`
## [1] &quot;ALG12&quot;
## 
## $`GO:0000010`
## [1] &quot;PDSS1&quot; &quot;PDSS2&quot;
## 
## $`GO:0000011`
## [1] &quot;RBSN&quot;
## 
## $`GO:0000012`
##  [1] &quot;APLF&quot;   &quot;APTX&quot;   &quot;E9PQ18&quot; &quot;LIG4&quot;   &quot;M0R2N6&quot; &quot;Q6ZNB5&quot; &quot;SIRT1&quot;  &quot;TDP1&quot;  
##  [9] &quot;TNP1&quot;   &quot;XRCC1&quot;</code></pre>
<p>It also has a function <code>get_GO_gene_sets</code> which is a specialisation of <code>get_ontological_gene_sets</code> for the Gene Ontology (GO) which can be called passing just a file path to the annotation file (official up-to-date version available at <a href="https://geneontology.org/" class="uri">https://geneontology.org/</a>).</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
